src/lint/lint.ts

```diff
@@ -0,0 +1,400 @@
+          });
+        }
+
+        if (rhs.kind === SyntaxKind.RightHandSide) {
+          (function noGrammarRestrictions(s: SymbolSpan | undefined) {
+            if (s === undefined) {
+              return;
+            }
+            if (s.symbol.kind === SyntaxKind.NoSymbolHereAssertion) {
+              let { line, column } = getLocationInGrammar(s.symbol.pos);
+              lintingErrors.push({
+                line,
+                column,
+                message: `Productions referenced in ${type}s should not include "no LineTerminator here" restrictions`,
+              });
+            }
+            // We could also enforce that lookahead restrictions are absent, but in some cases they actually do add clarity, so we just don't enforce it either way.
+
+            noGrammarRestrictions(s.next);
+          })(rhs.head);
+
+          if (rhs.constraints !== undefined) {
+            let { line, column } = getLocationInGrammar(rhs.constraints.pos);
+            lintingErrors.push({
+              line,
+              column,
+              message: `Productions referenced in ${type}s should not be gated on grammar parameters`,
+            });
+          }
+        }
+      }
+
+      // Filter out unused parameter errors for which the parameter is actually used in an SDO or Early Error
+      if (unusedParameterErrors.has(name)) {
+        let paramToError = unusedParameterErrors.get(name)!;
+        for (let [paramName, error] of paramToError) {
+          // This isn't the most elegant check, but it works.
+          if (rulesEles.some(r => r.innerHTML.indexOf('[' + paramName + ']') !== -1)) {
+            paramToError.delete(paramName);
+            // Yes, there's definitely big-O faster ways of doing this, but in practice this is probably faster for the sizes we will encounter.
+            lintingErrors = lintingErrors.filter(e => e !== error);
+          }
+        }
+      }
+    }
+  }
+
+  // *******************
+  // Enforce algorithm-specific linting rules
+
+  for (let algorithm of algorithms) {
+    let element = algorithm.element;
+    let location = getLocation(dom, element);
+
+    let reporter: Reporter = ({
+      line,
+      column,
+      message,
+    }: {
+      line: number;
+      column: number;
+      message: string;
+    }) => {
+      let trueLine = location.startTag.line + line - 1; // both jsdom and ecmarkdown have 1-based line numbers, so if we just add we are off by one
+      let trueCol = column;
+      if (line === 0) {
+        trueCol +=
+          location.startTag.col + (location.startTag.endOffset - location.startTag.startOffset);
+      } else {
+        trueCol += 1;
+      }
+      lintingErrors.push({ line: trueLine, column: trueCol, message });
+    };
+
+    let observer = composeObservers(...algorithmRules.map(f => f(reporter, element)));
+    let tree = parseAlgorithm(
+      sourceText.slice(location.startTag.endOffset, location.endTag.startOffset),
```

Comment:
Can't we just use like `element.innerHTML` here or something?

Reply:
`.innerHTML` is not guaranteed to give you the raw source - for example, it will translate `&#97;` in the original source into `a` in the `innerHTML`. Since we care about location information, that difference is unacceptable.